accepted/future-releases/nnbd/number-operation-typing.md

```diff
@@ -62,47 +62,80 @@ T add<T extends num>(T a, T b) => a + b;
+* Otherwise the static type of `e` is `num`.
 
-There are no special rules for `/` and `~/` because their return type is not `num`, and the return value's type is independent of the argument types. A `/` operation always returns a `double` and a `~/` operation always returns an `int`.
+With these typing rules, we cover all the instance members of `int` which has a return type of `num`, and we ensure that a using operands with the *same* type gives a result of that type, even if that type is a type variable (like `X extends int`) or promoted type variable (like `X & int`).
+
+There are no special rules for `/` and `~/` because their return type is not `num`, and the return value's type is independent of the argument types. A `/` operation always returns a `double` and a `~/` operation always returns an `int`.ma
+
+The rules for the binary operators can be summarized (non-normatively) as:
+
+
+|  *T* \\ *S*   | <: int | <: double | <: num  | dynamic |
+| :-----------: | ------ | --------- | ------- | ------- |
+|  **<: int**   | *T*    | double    | num     | num     |
+| **<: double** | *T*    | *T*       | *T*     | *T*     |
+|  **<: num**   | *T*    | double    | num/*T* | num     |
+
+where `<: num` here represents a subtype of `num` which is *not* also a subtype of `int` or `double`, and "num/*T*" covers the case where the static type is *T* when *S* <: *T* , and `num` when they are not.
+
+The rules are *not symmetric*. When possible, the type of the expression is the type of the first operand (which can be a type variable, even a promoted one), and otherwise it's a plain `int`, `double` or `num` type. The typing prefers the type of the first operand because it allows operations like `x += 1`  to work seamlessly.
+
+*In general, `x + integer` has the same static type as `x`.*
 
 ### Improved context type
 
 We extend type inference to take the special number rules into account.
 
-> For  `e1 + e2`, `e1 - e2`, `e1 * e2`, `e1 % e2` or `e1.remainder(e2)` where the static type of  `e1` is a non-`Never` subtype of `int`, if the context type of the entire expression is `int`, then the context type of `e2` is `int`. If the static type of `e1` is a non-`Never` subtype of `num` that is *not* a subtype of `double` and the context type of the entire expression is `double`, then the context type of `e2` is `double`.
+If `e` is an expression of the form  `e1 + e2`, `e1 - e2`, `e1 * e2`, `e1 % e2` or `e1.remainder(e2)`, where *C* is the context type of `e` and *T* is the static type of `e1`, and where both *C*  and *T* are non-`Never` subtypes of `num`, then:
+
+* If *C* <: `int` and *T* <: `int`, then the context type of `e2` is `int`. 
+* If *C* <: `double` and *T* is not a subtype of `double`, then the context type of `e2` is `double`.
+
+It is a compile-time error if the static type of `e2` is not a subtype of `num`. *(It is not necessarily a compile-time error if the static type of `e2` is not a subtype of _C_.)*
```

Comment:
Is this normative?  It seems like it's unnecessary, because it follows from the normal rules for analyzing method arguments.

Reply:
It can be either normative or not, since it states the same thing as the default rules. Except that it forgot about `dynamic`, so it's bad.
Let's remove the compile-time error requirement here, and just non-normatively state that the normal rules apply for assignability to the parameter type.

The question is what effect the context type has here.
It's not a requirement, just a hint.
What if the operand is `dynamic`, is it implicitly downcast to `num` or to *C*? 

It probably should be downcast to *C* because otherwise;

```dart
dynamic x = 1;
double x = 1 + x;  
```
would be invalid. `1+x` has type `num` unless `x` has type `double`, but with a context type of `double` on `x`, if we implicitly downcast it to `double`, then the code is allowed, and it will throw at the correct point.

Likewise:
```dart
dynamic x = 1;
int z = 1 + x;  // invalid unless we downcast `x` to `int`, then it is valid.
```

So, do we want to downcast `dynamic` to the context type. Is that what happens automatically?

If so, do we want to make it a compile-time error for:
```dart
int x = 1 + e;
```
if the static type of `e` is not `int`? Or just rely on the static type of `1 + e` not being `int` to get the error? The former would likely give the more precise error message.

WDYT? How much effect should the context type have:
* None, it's just a hint, not satisfying it won't matter.
* Cause downcast of `dynamic` to *C*.
* Cause compile-time error if static type is not *C*.
* Both of the above.

?

Reply:
> So, do we want to downcast `dynamic` to the context type.

I think we'd usually say that inference uses a context type, and the type of `x` is still `dynamic` when inference is complete. Then we compile `1 + x` to an invocation of `operator +` where it is checked dynamically that the value of `x` has type `num`. This is the standard treatment of expressions of type `dynamic` passed as actual arguments (or assigned to variables, etc.), so there's no need to re-specify it here.

It is possible for inference to fail, and we do report this as a compile-time error in some cases:

```dart
class A<X extends A<X>> {}
void main() => A();
```

If inference succeeds, I believe we can consider subsequent steps as independent of inference: Say, if `e` becomes `e1` during inference, then we just check the normal rules (like assignability) on `e1`, and it doesn't matter that the source code is actually `e` and `e1` is obtained by inference. A downcast in an initialization or assignment would be one of those subsequent steps, and we don't need to re-specify them.

In any case, I think the newest wording (where the paragraph shown as starting in line 116 above is all-commentary) is fine.

Reply:
ACK, so what we would need in order to support `int x = 1 + dyn;` is to say:

* If *T* <: `int` and *C* <: `int` and *S* is `dynamic`, then the static type of `e` is `int` and it's a *run-time error* if `e2` evaluates to a value which is not an `int`.
* If *C* <: `double` and *T* not <: `double`, then the static type of `e` is `double` and it's a run-time error if `e2` evaluates to a value which is not a `double`.

It would require combining the context and typing rules because then the typing rules depend on the context too.

We can do that. Should we?

Reply:
I want `int x = 1 + dyn;` to be a compile-time error in opted-in code.  It's an implicit downcast from `num` to `int`, and we don't allow implicit downcasts in opted-in code.  In my mind it's no different from something like this:
```dart
class C {
  Object m() => ...;
}
main() {
  C c = C().m(); // Error: implicit downcast from `Object` to `C`
}
```
Similarly, for `double x = 1 + dyn;`, it should be a compile time error, because it's an implicit cast from `num` to `double`.

I gues this means that the user won't be able to do `int x = ...; x += dyn;`, because that desugars to `x = x + dyn;`, and the static type of `x + dyn` is `num`.  I'm ok with that.

Reply:
I can see that this is compatible with how we treat `dynamic` otherwise.
The static type of the expression is `dynamic`, and that's assignable to any type. The fact that an argument is `dynamic` doesn't change the return type of a function call (unless inference is involved).

However, it is a little curious to me that we insert an implicit downcast, but we don't actually change the type of the expression. For `int foo(int x) => x; dynamic y = 1; foo(y);`, we say that the last invocation is equivalent to the explicit downcast `foo(y as int)`, but it isn't because the latter changes the static type of the expression.

We could instead say that the implicit cast inserted by the type system actually *changes* the static type of the expression, as if a real `as contextType` was inserted.

In that case, `int x = 1 + dyn;` would be equivalent to `int x = 1 + (dyn as int);`, and that code would work.

I hadn't though of `int x = ...; x += dyn;`, but if anything, that makes me want it to work. :)

Reply:
Ok, I see where you are coming from; you're right that `int x = 1 + (dyn as int);` would work, and it does seem reasonable to make `int x = 1 + dyn;` behave equivalently to that, because the only values for `dyn` that wouldn't result in a runtime error are ints.  That does a lot to ease my concerns.

I still feel a bit weird about it, though, and I can't decide if I'm just being stubborn or if there is a real user trap here.  Perhaps the crux of my concern is that in general, it's not safe to assume that a real `as contextType` gets inserted when there is a dynamic downcast, because when assigning to a promoted variable, we use a context type that's more specific than what's required, e.g.:
```dart
T f<T>(List<T> values) => values.first;
test(List<num> x) {
  dynamic y = <num>[];
  if (x is List<int>) {
    x = []; // Downward context is `List<int>` so this gets inferred as `<int>[]`
    x = y; // Ok; assignment de-promotes
  }
}
```

Reply:
I like the consistency of @stereotype441's approach: We use the special-cased context type to make `e2` an `int` when type inference can achieve that, but we don't try to make `e2` an `int` using any other means. When type inference does make `e2` an `int`, the developer can _trust_ this outcome.

This makes it natural for `1 + dyn` to apply the pure, signature based type check at `num` (turning it into `1 + (dyn as num)`), and not adding an implicit downcast all the way down to `int`. Consequently, as @stereotype441 mentioned, the initialization of `int x` has a further downcast from `num` to `int`, and that's an error with null-safety.

So it may seem tempting to "make it work in two more cases", but I think it's more important to keep the distinction between the safe steps (where type inference actually yields an expression that safely has the required type) and unsafe steps (where we implicitly add a dynamic type check). I think it's a better quality of service to ask developers to make those downcasts explicit.

Reply:
While I think overall the experience of inserting casts to the context type as deep as possible is probably better, I'm not sure it's a good idea to do so here when we don't do so in other places.  e.g. this code which is accepted and wouldn't be if we "pushed the downcast down":

```dart
void main() {
  dynamic d = "world";
  int x = (true) ? "hello" : d;
}
```

Reply:
As a counter-point to my own commentary, if we view this as an ad hoc use of overloading, then there are two things happening here:

1) A static overload resolution process which decides which version of (e.g.) `operator +` is being called
2) Normal inference + type checking.  

In this model, it makes sense that the cast would be on the second argument, since once you have determined the correct overload, the *type* of the overload is what really defines the context type for the argument, and just as with any other method invocation, we insert a cast to the parameter type if the inferred static type of the argument is not a subtype of the inferred (overload resolved) parameter type.

Reply:
I agree that I'm thinking of this in terms of overloading.

Now, if we ever get actual overloading, we might want these rules to fit into that.

If so, what we are seeing here is overloading based on context type (return type) and receiver. That might be a little far to go. Java does not have return type based overloading (they could, the [byte code allows methods that differ only in return type](https://community.oracle.com/docs/DOC-983207), but they don't base resolution on return type).

With only receiver+parameter based overloading, we can't choose between `double +(int, double)` and `int +(int, int)` when faced with `int+dynamic`.

(Then again, without self types, we can't handle the `Self operator+(int)`, which the rules here also specify, so we are already stretching overloading in order to make `x += 1` work for a any numeric type).

Reply:
Decision is that we do not make special rules for dynamic.